File: DONGUK/프로그래머스_WEEK 1/124나라의 숫자.py
def solution(n):
    nums = ['4', '1', '2']
    s = ''
    answer = 0
    while True:
        q = n // 3
        r = n % 3
        
        if r == 0:
            q = q - 1
            
        s = s + nums[r]
        
        n = q
        if q == 0:
            break
    
    answer = s[::-1]
    return answer

'''
1: 1
2: 2
3: 4 -> 1 0

4: 11 -> 1 1
5: 12
6: 14 -> 2 0

7: 21
8: 22
9: 24 -> 3 0

10: 41
11: 42
12: 44 -> 4 0

13: 111 4 1 ->  111
14: 112 4 2 -> 112
    
'''

File: JINSEOP/프로그래머스_지난_문제/124 나라의 숫자.cpp
#include <string>
#include <vector>
using namespace std;

void ternary(int n, string& answer) {
    while (n != 0) {
        if (n % 3 == 0) {
            answer = "4" + answer;
            n--;
        }
        else if (n % 3 == 2)
            answer = "2" + answer;
        else if (n % 3 == 1)
            answer = "1" + answer;

        n /= 3;
    }
}

string solution(int n) {
    string answer = "";

    ternary(n, answer);
    return answer;
}
